README.md
# Stream-client

[![Language C++](https://img.shields.io/badge/language-c++-blue.svg?logo=c%2B%2B)](https://isocpp.org)
[![Github releases](https://img.shields.io/github/release/TinkoffCreditSystems/stream-client.svg)](https://github.com/TinkoffCreditSystems/stream-client/releases)
[![Coverage Status](https://codecov.io/gh/Tinkoff/stream-client/branch/develop/graph/badge.svg?token=WN952FH80P)](https://codecov.io/gh/Tinkoff/stream-client)
[![License](https://img.shields.io/github/license/TinkoffCreditSystems/stream-client.svg)](./LICENSE)


This is a lightweight, header-only, Boost-based library providing client-side network primitives to easily organize and implement data transmission with remote endpoints.

This library:
* Inspired by and built around [Boost.Asio](http://boost.org/libs/asio/).
* Provides high-level constructs as [connector](#connector) and [connection pool](#connection-pool).
* Supports **TCP**/**UDP**/**SSL**/**HTTP**/**HTTPS** protocols.
* Uses [sockets](#sockets) with the same interface as boost::asio::ip ones.
* Allows transparent timeout/deadline control for all operations.
* Requires boost 1.65+ or 1.66+ if you are planning to use HTTP protocol.

## Why

If you are writing software on C++, which communicates with other services as a client, you probably already came-across with the problem - need to implement connectivity layer to provide network transport to all related services. This is exactly what this library's aim to solve.

## How to use

### Sockets

Sockets are implemented on top of `boost::asio::basic_socket` and provide classes with timeout control, so in most cases it's enough to just call *send()*/*receive()*/*write_some()*/*read_some()* with a deadline or timeout. Data supplied to I/O operations should be wrapped into `boost::asio::buffer`. Basically these clients are timeout-wrapped `boost::asio::ip` sockets and have the same interface.

Client streams classes:
* `stream_client::tcp_client` - plain TCP stream socket client. Supports *send()*/*rececive()* for admitted transfers of whole buffer along with *write_some()*/*read_some()* to transfer at least something.
* `stream_client::udp_client` - plain UDP socket client. Supports only *send()*/*receive()* without any acknowledgment as per UDP specs.
* `stream_client::ssl::ssl_client` - SSL-encrypted TCP client. Have the same functions as `tcp_client` plus SSL handshake and context control.
* `stream_client::http::http_client` - HTTP client. Wraps `tcp_client` with `boost::beast::http::parser` and `boost::beast::http::serializer` and have *perform()* function to make request-response calls.
* `stream_client::http::https_client` - HTTPS client. Same as `http_client` but uses `ssl_client` client underneath.

#### Example
```c++
const boost::asio::ip::tcp::endpoint endpoint(boost::asio::ip::address::from_string("127.0.0.1"), 12345);
const std::chrono::milliseconds connect_timeout(1000);
const std::chrono::milliseconds io_timeout(100);

// this makes connected socket and may take up to connect_timeout time
stream_client::tcp_client client(endpoint, connect_timeout, io_timeout);

// send() operation will transmit all contents of send_data or throw an error;
// also it will throw boost::asio::error::timed_out after io_timeout.
const std::string send_data = "test data";
client.send(boost::asio::buffer(send_data.data(), send_data.size()));

// receive() operation will read send_data.size() size or throw an error;
// also it will throw boost::asio::error::timed_out after io_timeout.
std::string recv_data(send_data.size(), '\0');
client.receive(boost::asio::buffer(&recv_data[0], send_data.size()));

// there are also non-throw overloads to these functions which
// will set boost::system::error_code instead of exception;
```

### DNS resolver

Resolver is implemented around the same idea - make sync, timed DNS resolver, therefore it uses `boost::asio::ip::basic_resolver` to perform actual resolution within specified timeouts.

It is the class template with a single parameter - result protocol, also you can specify which type of IP endpoints you are interested in - IPv4, IPv6 or both. Resolution performed with *resolve()* call which has a similar signature as sockets operations - you are free to specify timeout/deadline or use default one.

Resolver classes:
* `stream_client::resolver::tcp_resolver` - resolver which returns iterator of TCP endpoints.
* `stream_client::resolver::udp_resolver` - resolver which returns iterator of UDP endpoints.

#### Example
```c++
const std::chrono::milliseconds resolve_timeout(5000);

stream_client::resolver::tcp_resolver resolver("localhost", "12345",
                                               resolve_timeout,
                                               stream_client::resolver::ip_family::ipv4);

// resolve() guaranteed to return at least one endpoint or will throw an error;
// it will throw boost::asio::error::timed_out after resolve_timeout.
auto endpoints_it = resolver.resolve();

// returned iterator is an instance of boost::asio::ip::basic_resolver::iterator<Protocol>
```

### Connector

Connector uses [resolver](#dns-resolver) to perform DNS resolution and update the list of endpoints and uses them if requested with *new_session()*, to create a new socket of the specified protocol. The work of DNS resolver is wrapped into background thread which triggered upon creation or if there was an error in opening a new socket.

Target endpoint to open a new socket selected randomly from the DNS results, which makes it **balanced connector** to a remote host. Also, the connector allows specifying separate timeouts for resolution, connection, and I/O operation of new sockets.

Connector classes:
* `stream_client::connector::tcp_connector` - TCP connector, to obtain `stream_client::tcp_client` sockets.
* `stream_client::connector::udp_connector` - UDP connector, to obtain `stream_client::udp_client` sockets.
* `stream_client::connector::ssl_connector` - TCP connector, to obtain `stream_client::ssl::ssl_client` sockets.
* `stream_client::connector::http_connector` - HTTP connector, to obtain `stream_client::http::http_client` sockets.
* `stream_client::connector::https_connector` - HTTPS connector, to obtain `stream_client::http::https_client` sockets.

#### Example
```c++
const std::chrono::milliseconds resolve_timeout(5000);
const std::chrono::milliseconds connect_timeout(1000);
const std::chrono::milliseconds io_timeout(100);

// this will return immediately, starting background thread for name resolution
stream_client::connector::tcp_connector connector("localhost", "12345",
                                                  resolve_timeout, connect_timeout, io_timeout,
                                                  stream_client::resolver::ip_family::ipv4);

// this will acquire new socket or throw;
// it will throw boost::asio::error::timed_out after connect_timeout
std::unique_ptr<stream_client::tcp_client> client = connector.new_session();

// use the client
const std::string send_data = "test data";
std::string recv_data(send_data.size(), '\0');
client->send(boost::asio::buffer(send_data.data(), send_data.size()));
client->receive(boost::asio::buffer(&recv_data[0], send_data.size()));
```

### Connection pool

Represents container occupied with opened sockets. Uses [connector](#connector) to open new sockets in the background thread which is triggered once there are vacant places in the pool. User can call *get_session()* to obtain a socket from the pool and *return_session()* to give it back.

There are two strategies to refill the pool:
- **greedy** (`stream_client::connector::greedy_strategy`). If there are vacant places it will try to fill them with new sessions simultaneously. This is the default one.
- **conservative** (`stream_client::connector::conservative_strategy`). Will try to fill up to 2/3 of vacant places in the poll. If failed will back of for some time and retry later. Also, after failures it will create only one new session.

Both of them are defined in terms of `stream_client::connector::pool_strategy` interface, so you are free to implement new one.

Limitations:
1. Sockets that are already in the pool are not checked or maintained in any way. Hence, the pool doesn't guarantee that all sockets are opened at an arbitrary point in time due to the complexity of such checks for all supported protocols.
2. Nothing specific done with sockets upon their return within *return_session()*. Therefore, if they have or will have pending data to read, it will stay there until reading.

Considering this, the best strategy to use a connection pool is such:
1. Create it, specifying all timeouts as you need.
2. Once created, use *get_session()* to obtain opened socket.
3. Do needed I/O operations on the socket.
4. If 2-3 succeed, return it back with *return_session()*, else repeat point 2.

*It is important to discard sockets on failure and not reuse them, or request-response management will get nasty.*

Connection pools:
* `stream_client::connector::tcp_pool` - pool of `stream_client::tcp_client` sockets.
* `stream_client::connector::udp_pool` - pool of `stream_client::udp_client` sockets.
* `stream_client::connector::ssl_pool` - pool of `stream_client::ssl::ssl_client` sockets.
* `stream_client::connector::http_pool` - pool of `stream_client::http::http_client` sockets.
* `stream_client::connector::https_pool` - pool of `stream_client::http::https_client` sockets.

There are also aliases for the same pools but using conservative reconnection strategy:
* `stream_client::connector::tcp_conservative_pool`
* `stream_client::connector::udp_conservative_pool`
* `stream_client::connector::ssl_conservative_pool`
* `stream_client::connector::http_conservative_pool`
* `stream_client::connector::https_conservative_pool`

#### Example
```c++
const std::chrono::milliseconds resolve_timeout(5000);
const std::chrono::milliseconds connect_timeout(1000);
const std::chrono::milliseconds io_timeout(100);

auto connector_pool =
    std::make_shared<stream_client::connector::tcp_pool>("localhost", "12345",
                                                         resolve_timeout, connect_timeout, io_timeout,
                                                         stream_client::resolver::ip_family::ipv4);
const size_t threads_num = 10;
std::vector<std::thread> threads;
const std::string send_data = "test data";
std::mutex stdout_mutex;

for (size_t i = 0; i < threads_num; ++i) {
    threads.emplace_back([&send_data, &stdout_mutex, connector_pool]() {
        std::string recv_data(send_data.size(), '\0');

        auto client = connector_pool->get_session();
        // both these calls will throw exception on error or timeout
        client->send(boost::asio::buffer(send_data.data(), send_data.size()));
        client->receive(boost::asio::buffer(&recv_data[0], send_data.size()));
        // if there was no exception, return it
        connector_pool->return_session(std::move(client));

        const std::lock_guard<std::mutex> lk(stdout_mutex);
        std::cout << std::this_thread::get_id() << ": " << recv_data << std::endl;
    });
}

for (auto& t : threads) {
    t.join();
}
```

### Logging

The library uses basic logger interface implemented internally. You can modify logger level using `stream_client::set_log_level()` or `stream_client::get_log_level()` functions.

These levels are supported:
```cpp
enum class log_level : int
{
    trace = 0,
    debug,
    info,
    warning,
    error,
};
```

By default library prints messages to stdout with decent formatting. If you want to overwrite this behavior you can set you own logger via:
```cpp
void stream_client::set_logger(std::shared_ptr<stream_client::log_interface> logger);

void stream_client::set_logger(stream_client::log_level level, stream_client::log_func_type log_func);
```

Which allows to either overwrite logger instance of use a callback of with proper signature. These types are defines as:
```cpp
class log_interface
{
public:
    virtual void set_level(log_level level) noexcept = 0;
    virtual log_level get_level() const noexcept = 0;
    virtual void message(log_level level, const std::string& location, const std::string& message) const = 0;
};

using log_func_type = std::function<void(log_level level, const std::string& location, const std::string& message)>;
```

For more information please look inside [logger.hpp](./include/stream-client/logger.hpp).

## How to build

This library supposed to be somewhat multi-platform, however, it was tested and mainly used on ubuntu and macOS. </br>
Prefer [out-of-source](https://gitlab.kitware.com/cmake/community/-/wikis/FAQ#what-is-an-out-of-source-build) building.

### Ubuntu dependencies

```bash
sudo apt update
sudo apt install build-essential cmake libboost-dev libboost-system-dev libssl-dev
```

### macOS dependencies

```bash
brew install cmake pkg-config icu4c openssl boost
```

To build:
```bash
cmake -H. -Bbuild
cmake --build ./build
```

To install (sudo may be required):
```bash
cmake -H. -Bbuild -DSTREAMCLIENT_BUILD_TESTING=OFF -DSTREAMCLIENT_BUILD_DOCS=OFF -DSTREAMCLIENT_BUILD_EXAMPLES=OFF
cmake --build ./build --target install
```

Or test:
```bash
cmake -H. -Bbuild -DSTREAMCLIENT_BUILD_TESTING=ON -DSTREAMCLIENT_BUILD_DOCS=OFF -DSTREAMCLIENT_BUILD_EXAMPLES=OFF
cmake --build ./build
cmake -E chdir ./build ctest --output-on-failure
```

*All these commands assume you are in stream-client root folder*

### Cmake options

* **CMAKE_BUILD_TYPE** - [build type](https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_TYPE.html). `RelWithDebInfo` by default.
* **BUILD_SHARED_LIBS** - [build shared or static library](https://cmake.org/cmake/help/v3.0/variable/BUILD_SHARED_LIBS.html). `ON` by default.
* **STREAMCLIENT_BUILD_TESTING** - [build tests or not](https://cmake.org/cmake/help/latest/module/CTest.html). `OFF` by default
* **STREAMCLIENT_BUILD_EXAMPLES** - build library examples or not. `OFF` by default.
* **STREAMCLIENT_BUILD_DOCS** – build html (sphinx) reference docs. `OFF` by default.
* **OPENSSL_USE_STATIC_LIBS** - link statically or dynamically against found openssl. If *BUILD_SHARED_LIBS* is `OFF` then this options is set.
* **OPENSSL_ROOT_DIR** - folder where to look for openssl. Set by pkg-config of brew by default.

## License

Developed at **Tinkoff.ru** in 2020.\
Distibuted under **Apache License 2.0** [LICENSE](./LICENSE). You may also obtain this license at https://www.apache.org/licenses/LICENSE-2.0.

## Contacts

Author - i.s.vovk@tinkoff.ru\
Current maintainers - i.s.vovk@tinkoff.ru, n.suboch@tinkoff.ru


CHANGELOG.md
## 1.4.1 (2022-02-22)

### Misc

* improve logging for connection pool and conservative strategy
* fixed linkage for few translation units;


## 1.4.0 (2022-02-18)

### Features

* Implement logger facility and support user-defined overriding - (#19)


## 1.3.0 (2022-02-16)

### Features

* Support pool refill strategy and implement conservative and greedy (default) - (#20)

### Misc

* cmake: Use system googletest library if found
* cmake: Mute googletest warning
* workflow: Check .ipp files with clang-format


## 1.2.0 (2021-11-24)

### Features

* http/base_socket: Use dynamic allocator for transfered data


## 1.1.14 (2021-11-23)

### Bug Fixes

* connection_pool: Fix wrong error message "Too many open descriptors"


## 1.1.13 (2021-11-17)

### Bug Fixes

* connector: Fix data race for random generator
* tests/timeout: Remove time check for zero timeout

### Misc

* change googletest GIT_TAG to main
* workflow: Explicitly install openssl with brew
* readme: Add  n.suboch to maintainers list


## 1.1.12 (2021-09-21)

### Bug Fixes

* connector::base_connector: fix resolving thread spooling if attemts failed

### Misc

* Fix github 'lint' action on PRs


## 1.1.11 (2021-08-20)

### Bug Fixes

* connector::base_connector: fix UB in dealing with boolean flag

### Misc

* Add thread/address/UB sanitizers build to CI


## 1.1.10 (2021-08-17)

### Bug Fixes

* connector::base_connection_pool: fix UB in dealing with boolean flag

### Misc

* Add github actions
* Update README
* Lint the code


## 1.1.9 (2021-07-30)

### Features

* connector::base_connection_pool: add try_get_session method
* connector::base_connection_pool: change global lock in watch_pool_routine with several small locks 


## 1.1.8 (2021-07-14)

### Features

* ssl::stream_socket: set SNI hostname in ctor


## 1.1.7 (2021-06-30)

### Bug Fixes

* Fix mem leakage in static_allocator used for HTTP requests


## 1.1.6 (2021-06-29)

### Bug Fixes

* Fix static const undefined reference on some compilers for kHeaderLimit/kBodyLimit


## 1.1.5 (2021-02-12)

### Features

* Add idle timeout parameter for a pool. If session is sitting idle in the pool for a longer time it will replaced with new one


## 1.1.4 (2020-04-28)

### Bug Fixes

* TCP options not set in base_socket ctor


## 1.1.3 (2020-03-03)

### Bug Fixes

* get_session(timeout_or_deadline) return invalid socket


## 1.1.2 (2020-03-02)

### Features

* Add throwing get_session(timeout_or_deadline) function to the pool


## 1.1.1 (2020-03-02)

### Features

* Require standard openssl on macos (1.1+)
* Add Apache 2.0 LICENSE file

### Misc

* Small README fixes
* Style-related fixes


# 1.1.0 (2020-01-27)

### Features

* Apply clang-format as code style guard
* Add doxygen-format docs
* Add IPv4/IPV6/Any selector to stream_client::resolver::base_resolver<>
* Allow boost::asio::ip::basic_resolver_query::flags pass-through
* Separate TCP/UDP implementation since they are based on boost::asio::basic_stream_socket/basic_datagram_socket
* Add tests
* Add readme

### Bug Fixes

* Remove redundant ssl-handshake type parameter. Use client one


# 1.0.0 (2019-12-11)

- initial release


boost@1.66.rb
class BoostAT166 < Formula
  desc "Collection of portable C++ source libraries"
  homepage "https://www.boost.org/"
  url "https://dl.bintray.com/boostorg/release/1.66.0/source/boost_1_66_0.tar.bz2"
  sha256 "5721818253e6a0989583192f96782c4a98eb6204965316df9f5ad75819225ca9"
  head "https://github.com/boostorg/boost.git"

  bottle do
    cellar :any
    sha256 "78cb090c515e20aa7307c6619a055ffd8858cc6a3bd756958edbf34f463e4bc1" => :high_sierra
    sha256 "51e3b67625b8def53f5e5183a5ef98071e17b0b760dfc6baa65877d4528141ca" => :sierra
    sha256 "6352d4f65d7595c1eff62c6ce07588944122fa3305a77813fdbc16747b7dddce" => :el_capitan
  end

  option "with-icu4c", "Build regexp engine with icu support"
  option "without-single", "Disable building single-threading variant"
  option "without-static", "Disable building static library variant"

  deprecated_option "with-icu" => "with-icu4c"

  depends_on "icu4c" => :optional

  # Fix <experimental/string_view> for libc++7.0+
  # https://github.com/boostorg/asio/pull/91
  patch :DATA

  def install
    # Force boost to compile with the desired compiler
    open("user-config.jam", "a") do |file|
      file.write "using darwin : : #{ENV.cxx} ;\n"
    end

    # libdir should be set by --prefix but isn't
    bootstrap_args = ["--prefix=#{prefix}", "--libdir=#{lib}"]

    if build.with? "icu4c"
      icu4c_prefix = Formula["icu4c"].opt_prefix
      bootstrap_args << "--with-icu=#{icu4c_prefix}"
    else
      bootstrap_args << "--without-icu"
    end

    # Handle libraries that will not be built.
    without_libraries = ["python", "mpi"]

    # Boost.Log cannot be built using Apple GCC at the moment. Disabled
    # on such systems.
    without_libraries << "log" if ENV.compiler == :gcc

    bootstrap_args << "--without-libraries=#{without_libraries.join(",")}"

    # layout should be synchronized with boost-python and boost-mpi
    args = ["--prefix=#{prefix}",
            "--libdir=#{lib}",
            "-d2",
            "-j#{ENV.make_jobs}",
            "--layout=tagged",
            "--user-config=user-config.jam",
            "-sNO_LZMA=1",
            "install"]

    if build.with? "single"
      args << "threading=multi,single"
    else
      args << "threading=multi"
    end

    if build.with? "static"
      args << "link=shared,static"
    else
      args << "link=shared"
    end

    # Trunk starts using "clang++ -x c" to select C compiler which breaks C++11
    # handling using ENV.cxx11. Using "cxxflags" and "linkflags" still works.
    args << "cxxflags=-std=c++11"
    if ENV.compiler == :clang
      args << "cxxflags=-stdlib=libc++" << "linkflags=-stdlib=libc++"
    end

    system "./bootstrap.sh", *bootstrap_args
    system "./b2", "headers"
    system "./b2", *args
  end

  def caveats
    s = ""
    # ENV.compiler doesn't exist in caveats. Check library availability
    # instead.
    if Dir["#{lib}/libboost_log*"].empty?
      s += <<~EOS
        Building of Boost.Log is disabled because it requires newer GCC or Clang.
      EOS
    end

    s
  end

  test do
    (testpath/"test.cpp").write <<~EOS
      #include <boost/algorithm/string.hpp>
      #include <string>
      #include <vector>
      #include <assert.h>
      using namespace boost::algorithm;
      using namespace std;

      int main()
      {
        string str("a,b");
        vector<string> strVec;
        split(strVec, str, is_any_of(","));
        assert(strVec.size()==2);
        assert(strVec[0]=="a");
        assert(strVec[1]=="b");
        return 0;
      }
    EOS
    system ENV.cxx, "test.cpp", "-std=c++1y", "-L#{lib}", "-lboost_system", "-o", "test"
    system "./test"
  end
end

__END__
diff -Nur boost_1_66_0/boost/asio/detail/config.hpp boost_1_66_0-patched/boost/asio/detail/config.hpp
--- boost_1_66_0/boost/asio/detail/config.hpp   2017-12-06 00:00:00.000000000 -0700
+++ boost_1_66_0-patched/boost/asio/detail/config.hpp   2020-03-05 00:00:00.000000000 -0800
@@ -769,14 +769,21 @@
 #endif // !defined(BOOST_ASIO_HAS_STD_FUTURE)

 // Standard library support for experimental::string_view.
+// Note: libc++ 7.0 keeps <experimental/string_view> header but deprecates it
+// with an #error directive.
 #if !defined(BOOST_ASIO_HAS_STD_STRING_VIEW)
 # if !defined(BOOST_ASIO_DISABLE_STD_STRING_VIEW)
 #  if defined(__clang__)
 #   if (__cplusplus >= 201402)
 #    if __has_include(<experimental/string_view>)
-#     define BOOST_ASIO_HAS_STD_STRING_VIEW 1
-#     define BOOST_ASIO_HAS_STD_EXPERIMENTAL_STRING_VIEW 1
+#     if defined(_LIBCPP_LFTS_STRING_VIEW)
+#      define BOOST_ASIO_HAS_STD_EXPERIMENTAL_STRING_VIEW 1
+#      define BOOST_ASIO_HAS_STD_STRING_VIEW 1
+#     endif // defined(_LIBCPP_LFTS_STRING_VIEW)
 #    endif // __has_include(<experimental/string_view>)
+#    if __has_include(<string_view>)
+#     define BOOST_ASIO_HAS_STD_STRING_VIEW 1
+#    endif // __has_include(<string_view>)
 #   endif // (__cplusplus >= 201402)
 #  endif // defined(__clang__)
 #  if defined(__GNUC__)
